_posts/2019-09-05-kafka-retry-dlq.md

```diff
@@ -0,0 +1,194 @@
+ 
+### Nieskończone ponawianie w miejscu
+![2019-09-05-Kafka-DLQ-Strategy02.png](/assets/img/posts/2019-09-05-kafka-retry-dlq/2019-09-05-Kafka-DLQ-Strategy02.png)
+Gdy nie możemy pozwolić sobie na utratę komunikatów, najprostszym podejściem jest ponawianie do skutku. Oczywistą konsekwencją jest tzw. zatrzymanie świata. Dopóki błąd nie zostanie poprawiony albo zewnętrzny system udrożniony - żaden kolejny komunikat nie zostanie przetworzony. Takie rozwiązanie jest konieczne w przypadku, gdy chcemy zachować kolejność przetwarzania zdarzeń systemie. W ten scenariusz tym bardziej wpisuje się potrzeba stałego monitoringu.
+ 
+### Skończone ponawianie w miejscu z topikiem błędów
+![2019-09-05-Kafka-DLQ-Strategy03.png](/assets/img/posts/2019-09-05-kafka-retry-dlq/2019-09-05-Kafka-DLQ-Strategy03.png) 
+Wyobraźmy sobie teraz co się stanie, jak nieco poluzujemy wymaganie bezwzględnego zachowania kolejności. Załóżmy, że próbujemy przez jakiś czas ponawiać, ponieważ statystyka i doświadczenie podpowiada nam, że 99% problemów z przetwarzaniem komunikatów jest chwilowych i samoczynnie ustępuje po pewnym czasie. Dodatkowo komunikaty, których nie udało się przetworzyć, **kopiujemy** na oddzielny topik traktowany jako DLQ. Dzięki temu mamy od razu wyłuskane problematyczne wiadomości i możemy uruchomić na nich osobną grupę konsumentów. 
+
+Krótkie wyjaśnienie, dlaczego komunikaty są **kopiowane** a nie przenoszone. Odpowiedź jest bardzo prosta - nie mogą być przenoszone. Wynika to z kolejnego fundamentu architektonicznego Kafki czyli niezmienności topików (ang. topics immutability). Niezależnie jaka była przyczyna błędu, komunikat na zawsze pozostanie utrwalony. Istnieją sposoby na radzenie sobie z tym tematem i wrócimy do tego później. 
+ 
+### Skończone ponawianie na wydzielonym topiku
+![2019-09-05-Kafka-DLQ-Strategy04.png](/assets/img/posts/2019-09-05-kafka-retry-dlq/2019-09-05-Kafka-DLQ-Strategy04.png)
+Dochodzimy niniejszym do naszego ostatecznego rozwiązania. Skoro mamy osobny topik dla zepsutych wiadomości to może warto wprowadzić kolejny, na którym odbywa się ponawianie. W tym modelu jeszcze bardziej luzujemy konieczność zachowania kolejności ale dostajemy w zamian możliwość bezprzerwowego przetwarzania głównego topiku. Wiadomości kaskadowo kopiowane są najpierw na topik wiadomości ponawianych a w przypadku niepowodzenia - topik DLQ (technicznie powinniśmy nazwać go DLT, ale zostańmy przy akronimie DLQ, jako że jest on dobrze kojarzony z tego rodzaju technikami). 
```

Comment:
Rozwinąłbym trochę temat luzowania zachowania kolejności, bo wydaje mi się to mało namacalne tutaj. Coś o zależności kolejnych eventów od siebie, jak to może rozspójnić system, albo wprost przeciwnie - kiedy eventy od siebie nie zależą, i jaka wypływa z tego wolność w obsłudze błędów ;)

Reply:
dopisałem szersze wyjaśnienie do Strategii 3, bo tam pojawiasie po raz pierwszy wzmianka o luzowaniu